test/Test/Cardano/Prelude/Helpers.hs

```diff
@@ -9,21 +9,30 @@ where
 
 import Cardano.Prelude
 
+import Data.Data (Data, toConstr)
+import Data.Function (on)
 import Formatting (Buildable, build, sformat)
+import GHC.Stack (HasCallStack, withFrozenCallStack)
 
 import Hedgehog (MonadTest, success, (===))
 import Hedgehog.Internal.Property (failWith)
 
 assertEitherIsLeft
-  :: (MonadTest m, Buildable c) => (a -> Either b c) -> a -> m ()
-assertEitherIsLeft func val = case func val of
-  Left  _   -> success
+  :: (Buildable b, Buildable c, Data b, HasCallStack, MonadTest m)
+  => (a -> Either b c)
+  -> a
+  -> b
+  -> m ()
```

Comment:
I find this type a little counterintuitive, because it appears like it would literally compare the values of type `b` (the provided value, and the failure which goes into the `Left`), but it actually doesn't - it just checks that the constructors are equal.

If we redefined this at type
```
assertEitherIsLeft
  :: (Buildable b, Buildable c, Data b, HasCallStack, MonadTest m)
  => (a -> Either b c)
  -> a
  -> Constr
  -> m ()
```
and compared the `Constr`-typed value with `toConstr` of the value in the `Left`, then the type would capture the functionality more accurately.

The disadvantage of this, though, is that the client of these function has to call `toConstr` on their datatype. i.e. the abstraction is leaky.

Reply:
My issue with this is a little deeper (and reflected in the fact that I just added `asserIsJust` and `assertIsNothing`.

I don't understand why the assertion as it is requires a function as the first parameter. Would it not be better to just assert that the value is as expected? Eg:
```
assertEitherIsLeft
  :: (MonadTest m, Buildable b) => Either a b -> m ()
assertEitherIsLeft val = case func val of
  Left  _   -> success
  Right res -> failWith Nothing (show $ sformat build res)
```
Then we could also drop `Either` from the name so its just `assertIsLeft` and `assertIsRIght`.

By all means try this out to make sure they still do what you need, but I would be curious if this simplified version does not work out.

Reply:
@erikd Yep this is a sensible change.